util/codec/codec_test.go

```diff
@@ -1216,3 +1216,96 @@ func (s *testCodecSuite) TestHashChunkColumns(c *C) {
 		c.Assert(vecHash[2].Sum64(), Equals, rowHash[2].Sum64())
 	}
 }
+
+func prepareCollationData() (int, *chunk.Chunk, *chunk.Chunk) {
+	tp := types.NewFieldType(mysql.TypeString)
+	tps := []*types.FieldType{tp}
+	chk1 := chunk.New(tps, 3, 3)
+	chk2 := chunk.New(tps, 3, 3)
+	chk1.Reset()
+	chk2.Reset()
+	chk1.Column(0).AppendString("aaa")
+	chk2.Column(0).AppendString("AAA")
+	chk1.Column(0).AppendString("😜")
+	chk2.Column(0).AppendString("😃")
+	chk1.Column(0).AppendString("À")
+	chk2.Column(0).AppendString("A")
+	return 3, chk1, chk2
+}
+
+func (s *testCodecSuite) TestHashGroupKeyCollation(c *C) {
+	sc := &stmtctx.StatementContext{TimeZone: time.Local}
+	tp := types.NewFieldType(mysql.TypeString)
+	n, chk1, chk2 := prepareCollationData()
+
+	tp.Collate = "utf8_general_ci"
+	buf1 := make([][]byte, n)
+	buf2 := make([][]byte, n)
+	buf1, err := HashGroupKey(sc, n, chk1.Column(0), buf1, tp)
+	c.Assert(err, IsNil)
+	buf2, err = HashGroupKey(sc, n, chk2.Column(0), buf2, tp)
+	c.Assert(err, IsNil)
+	for i := 0; i < n; i++ {
+		c.Assert(len(buf1[i]), Equals, len(buf2[i]))
+		for j := range buf1 {
+			c.Assert(buf1[i][j], Equals, buf2[i][j])
+		}
+	}
+}
+
+func (s *testCodecSuite) TestHashChunkRowCollation(c *C) {
+	sc := &stmtctx.StatementContext{TimeZone: time.Local}
+	tp := types.NewFieldType(mysql.TypeString)
+	tps := []*types.FieldType{tp}
+	n, chk1, chk2 := prepareCollationData()
+	cols := []int{0}
+	buf := make([]byte, 1)
+
+	tp.Collate = "bin"
+	for i := 0; i < n; i++ {
+		h1 := crc32.NewIEEE()
+		h2 := crc32.NewIEEE()
+		c.Assert(HashChunkRow(sc, h1, chk1.GetRow(i), tps, cols, buf), IsNil)
+		c.Assert(HashChunkRow(sc, h2, chk2.GetRow(i), tps, cols, buf), IsNil)
+		c.Assert(h1.Sum32(), Not(Equals), h2.Sum32())
+		h1.Reset()
+		h2.Reset()
+	}
+
+	tp.Collate = "utf8_general_ci"
```

Comment:
The current problem is, for now we have to call `collate.SetNewCollationEnabled(true)` to get the "real" general_ci collator. Maybe we should add an interface for the unit tests.

Reply:
I add the call `collate.SetNewCollationEnabledForTest(true)` in each tests about collations now.
And all these tests are processed by the `testCollationSuite`, which is a `SerialSuites`.